devito/arguments.py

```diff
@@ -158,198 +145,398 @@ class PtrArgument(Argument):
 
     is_PtrArgument = True
 
-    def __init__(self, name, provider):
-        super(PtrArgument, self).__init__(name, provider, provider.value)
-
-    def verify(self, value):
-        self._value = value or self._value
-        return True
-
-    def reset(self):
-        if callable(self.default_value):
-            self._value = self.default_value()
-        else:
-            self._value = self.default_value
+    def __init__(self, provider):
+        super(PtrArgument, self).__init__(provider.name,
+                                          [ValueDependency(provider)], provider.dtype)
 
 
-class ArgumentProvider(object):
-
-    """ Abstract base class for any object that, post code-generation, might resolve
-        resolve to runtime arguments. We assume that one source object (e.g. Dimension,
-        SymbolicData) might provide multiple runtime arguments.
+class ArgumentEngine(object):
+    """ Class that encapsulates the argument derivation and verification subsystem
     """
+    def __init__(self, stencils, parameters, dle_arguments):
+        self.parameters = parameters
+        self.dle_arguments = dle_arguments
+        argument_mapper = self._build_argument_mapper(parameters)
```

Comment:
I don't like the suffix ``_mapper``, as this seems to be a list. Is this a list?

Reply:
Fixed